Module to handle rotary decoders

pjsg opened the pull request:
This is the PR that goes with the discussion in #1058 

This makes use of the new task infrastructure to send events from interrupt land down to the lua level. I ended up using software timers for the long delays (the 'long' press time and the double click time).

It works for me reliably. It is not difficult to overflow the queue, but this ends up just losing `TURN` events -- the absolute position is still tracked. 

I added a simple hooking mechanism to the GPIO interrupt handler so that other modules can register and get interrupts passed to them. This module is the first user of that feature.

marcelstoer commented:
I restarted the build after I merged #1070 - all green now.

Pls add the module to https://github.com/nodemcu/nodemcu-firmware/wiki/Module-list once this PR is merged.

ABonner commented on docs/en/modules/rotary.md:
```diff
@@ -0,0 +1,129 @@
+
+#### Parameters
+- `channel` The rotary module supports three switches. The channel is either 0, 1 or 2.
+- `eventtype` This defines the type of event being registered. This is the logical or of one or more of `PRESS`, `RELEASE` or `TURN`.
```
Also `CLICK` and `DBLCLICK`?

dfeinzeig replied:
bump

pjsg commented:
I back-merged dev (and fixed the conflicts). I also squashed down into a single commit.

jmattsson commented:
I'm going to leave this for @TerryE to sign off on, seeing as he's fully across the GPIO interrupts at the moment :)

Looks nifty though!

TerryE commented:
Sorry Philip, I have just pulled your module and will go through the code before either feeding back comments or doing the merge.

TerryE commented:
Philip my bad and my apologies for not chipping in sooner.  I left this to you in the earlier discussions, since I was fighting other fires, but the issue of the GPIO platform changes is one of wider scope.  if you want, I can push a version to my Repo containing my suggested changes so we can discuss,

pjsg commented:
I'll certainly take a look at your branch....

TerryE commented:
Philip, you have design authority over the rotary module. However,  the GPIO driver interface is  matter of consensus, so we need to be quorate on whatever approach we finally agree on.

TerryE commented:
@pjsg Philip, here is a gist of the diff,  I haven't bothered to push it up as a new branch.  This is as about lightwieght as I can get it at runtime but it still adds ~500 bytes of code, so I've also added an extra config option to suppress this if you don't need it. I am overdue bed.

pjsg commented:
I don't see a gist.....

TerryE commented:
Sorry [here](https://gist.github.com/TerryE/742b93ae58c3d9f40737)

TerryE commented:
And thinking about it using 2 mallocs is daft. Doing one of n * (sizeof( void *)+sizeof(uint32)) makes the code simpler and its size smaller.

pjsg commented:
I fixed it to use a single malloc in the register function. I also added a way to update the set of bits for a hook -- this allows you to do an atomic update of moving from one set of bits to another.

It still works!!

pjsg commented:
The reason to be able to change the registered bits is when a driver wants to hook more bits -- for example, if the user wants to open a second rotary switch. You can't unregister the first switch and then re-register both (as you may lose interrupts). You can't do this under a disabled interrupt (memory allocation in this state is bad). You don't want to register the same hook multiple times. 

I'll reorganize the code a bit.....

pjsg commented:
The code does have a check that a valid combination of LUA_USE_MODULES_ROTARY and the two GPIO interrupts are set. The code gets compiled anyway, and doesn't cause any link problems.

As the number of modules grows, I wonder if it is worth adding big #ifdef LUA_USE_MODULES_blah / #endif surrounding each of the source files? This might speed up the build process (or at least keep it constant).

TerryE commented:
Philip, I'll merge this because whilst we may decided to evolve this through a couple more cycles, this is a stable enough baseline to moved to dev.
